### ISSUE TITLE
[MPS] index_select / F.embedding silently return all-zero rows when the source tensor exceeds 2**31 elements (fixed on main, still broken in release/2.13)

### ISSUE THREAD

**Issue by KitPan** [pytorch/pytorch#188756]
### 🐛 Describe the bug

On MPS, `F.embedding` / `nn.Embedding` / `torch.index_select` silently return **all-zero rows** (no error, no NaN) for every index `i >= ceil(2**31 / dim)` when the source weight tensor has more than `2**31` elements. Indices below that boundary return bit-exact rows, so the corruption is partial and easy to miss.

This is a **silent correctness** bug in the current stable release (2.12.0). It is already fixed on `main` by the native-Metal-kernel migration in #187109 (verified on `2.14.0.dev20260701`), but as of today the **`release/2.13` branch still contains the broken MPSGraph path**, so 2.13 would ship with it (see "Status" below).

#### Minimal repro (no model download; ~10.5 GB unified memory)

```python
import torch
import torch.nn.functional as F

rows, dim = 262144, 10752  # 2,818,572,288 elements > 2**31 (google/gemma-4-E4B-it's PLE table)
w = torch.ones(rows, dim, dtype=torch.bfloat16).to("mps")

first_bad = -(-(2**31) // dim)  # ceil(2**31 / dim) = 199729
ids = torch.tensor([0, first_bad - 1, first_bad, rows - 1], device="mps")

print(F.embedding(ids, w).float().sum(1).cpu())            # every row sums to dim if correct
print(torch.index_select(w, 0, ids).float().sum(1).cpu())
print(w[ids].float().sum(1).cpu())                         # advanced indexing, for comparison
```

Output on **torch 2.12.0** (macOS 26.5.1, M5 Max):

```text
tensor([10752., 10752.,     0.,     0.])   # F.embedding      <- rows 199729 and 262143 are all-zero
tensor([10752., 10752.,     0.,     0.])   # index_select     <- same
tensor([10752., 10752., 10752., 10752.])   # w[ids] advanced indexing is correct
```

Output on **torch 2.14.0.dev20260701** (same machine): all three lines are `tensor([10752., 10752., 10752., 10752.])` — fixed.

CPU produces the correct result for all three ops.

#### Observations (from a fuller bisecting repro, happy to share)

- The first wrong id is exactly `ceil(2**31 / dim)` = 199729, binary-searched: id 199728 is bit-exact vs CPU, id 199729 is zeros. That is precisely the first row whose flat start offset `i * dim` exceeds `INT32_MAX`.
- Wrong rows are entirely zero — not wrapped/shifted data (checked against a table where every row carries a nonzero row-index marker).
- Row 199728, whose *tail* elements live at flat offsets past `INT32_MAX` (the row spans elements `[2,147,475,456 .. 2,147,486,207]`), reads back correctly. Only rows whose *start* offset exceeds `INT32_MAX` fail — so it is the `index * slice_stride` product that overflows, not per-element addressing.
- Single-row slicing `w_mps[i]` and advanced indexing `w_mps[ids]` are both correct for all ids (those lower to the native Metal index kernels, which use 64-bit offsets), and the table round-trips bit-exact through slicing — the device copy is intact; only the gather is broken.
- All indices involved are valid; this is not the OOB-index case of #154235 / #144824, though the failure surface is the same "OOB reads return 0" behavior.

#### Root cause (as far as I can tell)

In 2.12.0, `index_select_out_mps` lowers to MPSGraph `gatherWithUpdatesTensor:indicesTensor:axis:` with no tensor-size guard:
https://github.com/pytorch/pytorch/blob/v2.12.0/aten/src/ATen/native/mps/operations/Indexing.mm#L672-L676

MPSGraph's gather appears to do signed 32-bit offset arithmetic internally; for rows starting past `INT32_MAX` the overflowed offset presumably lands out of bounds, and MPS's OOB-reads-return-0 behavior (#144824) turns that into silent zeros. `F.embedding` takes the same path because dense embedding forward lowers to `index_select`.

#### Real-world impact

This is how we found it: `google/gemma-4-E4B-it`'s per-layer-embedding table is `nn.Embedding(262144, 10752)` — 2.82e9 elements in bf16. On MPS, every token id ≥ 199729 silently gets a zero embedding row. ASCII/European token ids sit below the boundary while CJK token ids sit above it, so the model runs perfectly in English and produces degraded Chinese/Japanese/Korean — it masquerades as "the model is bad at CJK" rather than a backend correctness bug. Any vocab/table crossing 2**31 elements is affected the same way.

#### Status and asks

- `main`: **fixed** by #187109 (merged 2026-06-15), which replaced the MPSGraph gather with a native Metal kernel using `long` offsets. Verified fixed on nightly `2.14.0.dev20260701`.
- `release/2.13`: **still broken** — the branch still has the MPSGraph `gatherWithUpdatesTensor:` path: https://github.com/pytorch/pytorch/blob/release/2.13/aten/src/ATen/native/mps/operations/Indexing.mm#L721

Given that, this issue is mainly three asks:

1. **Cherry-pick the `index_select` part of #187109 (or add a `numel() > INT32_MAX` guard/fallback) into `release/2.13`**, so 2.13 does not ship silent data corruption for large embedding tables. A loud `TORCH_CHECK` would already be a big improvement over silent zeros.
2. **Add a `> 2**31`-element regression test for `index_select`/`F.embedding`** so the MPSGraph-era behavior can't quietly return. I don't see one in `test/test_mps.py` (there are analogous large-tensor regression tests, e.g. `test_rand_4b` and an `index_copy_` one; note also #182848 — `@largeTensorTest` used to skip silently on MPS). The repro above fits the `test_rand_4b` pattern (guard on `recommended_max_memory()`).
3. **Audit remaining MPSGraph gather/scatter users for the same 32-bit limit** — #187109 notes `flip` is the only remaining MPSGraph user in `Indexing.mm`, but other files still lower gather/scatter-like ops to MPSGraph.

Related: #84520 (MPS 32-bit limits umbrella — but here nothing asserts, it corrupts silently), #144824 (OOB indexing returns 0), #154235 (missing index range-check on MPS), #182052 and #185663 (same class: 32-bit offset overflow silently corrupting large-tensor ops; both fixed).

Suggested labels: `module: mps`, `module: correctness (silent)`.

### Versions

<details>

```text
PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A
OS: macOS 26.5.1 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: version 4.3.4
Libc version: N/A
Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:07:20) [Clang 20.1.8 ] (64-bit runtime)
Python platform: macOS-26.5.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
Is XPU available: False
Is XNNPACK available: True
CPU: Apple M5 Max
Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
```

Also tested: `torch==2.14.0.dev20260701` (nightly, same machine) — not affected (fixed).

</details>

cc @pragupta @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20 @albanD

**Comment by malfet** [comment]
cc: @atalman in case cherry-pick window is still open

**Comment by atalman** [comment]
Hi @malfet please see: https://github.com/pytorch/pytorch/issues/186934#issuecomment-4844225377 we are currently in final testing phase for release 2.13 the cherry-pick windows is closed

**Comment by malfet** [comment]
Closing, as it's fixed in 2.14.0
